internal/controllers/clusterextension_controller.go

```diff
@@ -202,21 +179,110 @@ func (r *ClusterExtensionReconciler) reconcile(ctx context.Context, ext *ocv1alp
 	return ctrl.Result{}, nil
 }
 
-func (r *ClusterExtensionReconciler) variables(ctx context.Context) ([]deppy.Variable, error) {
+func (r *ClusterExtensionReconciler) resolve(ctx context.Context, ext *ocv1alpha1.ClusterExtension) (*catalogmetadata.Bundle, error) {
 	allBundles, err := r.BundleProvider.Bundles(ctx)
 	if err != nil {
 		return nil, err
 	}
-	clusterExtensionList := ocv1alpha1.ClusterExtensionList{}
-	if err := r.Client.List(ctx, &clusterExtensionList); err != nil {
-		return nil, err
+
+	var successors successorsPredicateFunc = legacySemanticsSuccessorsPredicate
+	if features.OperatorControllerFeatureGate.Enabled(features.ForceSemverUpgradeConstraints) {
+		successors = semverSuccessorsPredicate
 	}
-	bundleDeploymentList := rukpakv1alpha2.BundleDeploymentList{}
-	if err := r.Client.List(ctx, &bundleDeploymentList); err != nil {
-		return nil, err
+
+	packageName := ext.Spec.PackageName
+	channelName := ext.Spec.Channel
+	versionRange := ext.Spec.Version
+
+	predicates := []catalogfilter.Predicate[catalogmetadata.Bundle]{
+		catalogfilter.WithPackageName(packageName),
+	}
+
+	if channelName != "" {
+		predicates = append(predicates, catalogfilter.InChannel(channelName))
+	}
+
+	if versionRange != "" {
+		vr, err := mmsemver.NewConstraint(versionRange)
+		if err != nil {
+			return nil, fmt.Errorf("invalid version range %q: %w", versionRange, err)
+		}
+		predicates = append(predicates, catalogfilter.InMastermindsSemverRange(vr))
+	}
+
+	// Do not include bundle versions older than currently installed unless UpgradeConstraintPolicy = 'Ignore'
+	var installedVersion string
+	if ext.Spec.UpgradeConstraintPolicy != ocv1alpha1.UpgradeConstraintPolicyIgnore {
+		bd := &rukpakv1alpha2.BundleDeployment{}
+		err := r.Client.Get(ctx, types.NamespacedName{Name: ext.GetName()}, bd)
+		if client.IgnoreNotFound(err) != nil {
+			return nil, err
+		}
+
+		if !apierrors.IsNotFound(err) && bd.Spec.Source.Image != nil && bd.Spec.Source.Image.Ref != "" {
+			bundleImage := bd.Spec.Source.Image.Ref
+			// find corresponding bundle for the installed content
+			installedBundleResultSet := catalogfilter.Filter(allBundles, catalogfilter.And(
+				catalogfilter.WithPackageName(ext.Spec.PackageName),
+				catalogfilter.WithBundleImage(bundleImage),
+			))
+			if len(installedBundleResultSet) == 0 {
+				return nil, fmt.Errorf("bundle with image %q for package %q not found in available catalogs but is currently installed via BundleDeployment %q", bundleImage, ext.Spec.PackageName, bd.Name)
+			}
+
+			sort.SliceStable(installedBundleResultSet, func(i, j int) bool {
+				return catalogsort.ByVersion(installedBundleResultSet[i], installedBundleResultSet[j])
+			})
+			installedBundle := installedBundleResultSet[0]
+			installedBundleVersion, err := installedBundle.Version()
+			if err != nil {
+				return nil, err
+			}
+			installedVersion = installedBundleVersion.String()
```

Comment:
Can we have this in a separate method, that returns the installedVersion, so that its easier to change when we have a helm integration?

Also, to understand this better - To find the installed version in the cluster, we are looking at the spec of the bundle deployment. Since the spec is the specified (desired) component rather than what exists on cluster, do we need to look at the status rather?

There was a similar discussion related to this for the extension controller, and since we can't read from our own status, it was suggested to add an annotation containing the installed bundle Version into the object. Which is what is implemented [here](https://github.com/operator-framework/operator-controller/blob/25af3b02ab8e2737ec982628dccdaca38f0dc618/internal/controllers/extension_controller.go#L468) and is discussed in [slack](https://redhat-internal.slack.com/archives/C06KP34REFJ/p1710352584410229).

Reply:
> Can we have this in a separate method, that returns the installedVersion, so that its easier to change when we have a helm integration?

Yes, that's the plan for the final version.

> Also, to understand this better - To find the installed version in the cluster, we are looking at the spec of the bundle deployment. Since the spec is the specified (desired) component rather than what exists on cluster, do we need to look at the status rather?

Your understanding is correct. We had a similar discussion when we were initially implementing this and I can't remember why we decided to do it this way. I also can't find the relevant thread anymore (was a while ago on Kube slack).

But I think it is ok in this context: here were are reconciling `ClusterExtension` and are reading from `BundleDeployment`'s spec. The actual cluster state is not important here as we are trying to retrieve the bundle version from the previous reconciliation (e.g. previous desired state).

I suggest that for now we do the same approach as in main (based on spec) to reduce the blast radius. I also expect this bit to change once we start accounting for non-image content.

Reply:
Fair, since this is based on BD and we are going to modify it soon it's better we don't spend much time on this. I'm good to merge with what we have now.